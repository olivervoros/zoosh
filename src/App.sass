$spinner-color: black
$spinner-accent: gray
$spinner-size: 40px
@import "~scss-spinners/spinners"

@mixin border-radius($pixel...)
  border-radius: $pixel

body
  background-color: #e2ebef
  font-family: 'Trocchi',sans-serif

.break
  flex-basis: 100%
  height: 0

.button
  border: none
  padding: 15px 32px
  margin: 0 20px
  text-align: center
  text-decoration: none
  display: inline-block
  font-size: 16px
  background-color: coral
  @include border-radius(5px)
  cursor: pointer
  text-transform: uppercase
  box-shadow: none
  color: white
  &:hover
    color: white
    background-color: #4CAF50

.resetButton
  @extend .button
  background-color: #98ff98
  color: black

input[type=text]
  width: 40%
  padding: 12px 20px
  margin: 8px 0
  box-sizing: border-box


.container
  margin-right: auto
  margin-left: auto
  max-width: 960px
  padding-right: 10px
  padding-left: 10px
  text-align: center

.resultCard
  background-color: white
  @include border-radius(10px)
  padding: 10px
  margin: 30px 0px
  display: flex
  flex-wrap: wrap
  h3
   color: black

.resultData
  width: 80%

.resultPoster
    img
      max-width: 100px
      width: 100px
      height: auto


.wikiArticle
  a
    color: black

.spinner
  text-align: center

.wikiArticle p.wikiArticleBody
  text-align: center
  padding: 20px
  margin: 20px
  line-height: 2

.wikiRelatedArticle p a
  text-align: center
  padding: 20px
  margin: 20px
  line-height: 2
  color: black


