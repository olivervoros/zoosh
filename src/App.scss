$spinner-color: black;
$spinner-accent: gray;
$spinner-size: 40px;
@import "~scss-spinners/spinners";

.container
  margin-right: auto
  margin-left: auto
  max-width: 960px
  padding-right: 10px
  padding-left: 10px