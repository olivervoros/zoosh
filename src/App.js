import React, {Component} from 'react';
import SearchMovieForm from "./components/SearchMovieForm";
import './App.scss';

class App extends Component {

    render() {

        return (
            <div className="container">
                <p>Welcome to the Zoosh movie database searcher.</p>
                <SearchMovieForm></SearchMovieForm>
            </div>
        )
    }
}

export default App;