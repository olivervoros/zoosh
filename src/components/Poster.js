import React from 'react'
import '../App.sass';
import {imageExists, CHECK_IF_IMAGE_EXISTS} from "../Helper";

function Poster(props) {

        return (
            typeof props.poster !== 'undefined' && CHECK_IF_IMAGE_EXISTS && imageExists(props.poster)  &&
            <div className="resultPoster">
                <p><img alt="" src={props.poster} /></p>
            </div>
        )

}

export default Poster;