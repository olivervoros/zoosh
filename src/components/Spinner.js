import React, {Component} from 'react'
import '../App.scss';

class Spinner extends Component {

    render() {

        return (
            <div className="spinner round"></div>
        )
    }

}

export default Spinner;