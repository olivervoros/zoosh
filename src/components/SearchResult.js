import React, {Component} from 'react'
import '../App.sass';
import axios from "axios";
import { WIKIPEDIA_OPENSEARCH_API_ENDPOINT, WIKIPEDIA_GET_LEAD_ARTICLE_ENDPOINT, WIKIPEDIA_GET_RELATED_ARTICLES_ENDPOINT } from "../Helper";
import SearchResultWikipedia from "./SearchResultWikipedia";
import Poster from "./Poster";
import Spinner from "./Spinner";

class SearchResult extends Component {

    constructor(props) {
        super(props);
        this.state = { searchResultWikipedia: [], isLoading: false, displayError: false };
    }

    async searchWikipedia(event, searchTitle, imdbID) {
        event.preventDefault();
        await this.setState({ isLoading: true });
        try {
            let result = await axios.get(WIKIPEDIA_OPENSEARCH_API_ENDPOINT + searchTitle);
            let leadArticleResult = await axios.get(WIKIPEDIA_GET_LEAD_ARTICLE_ENDPOINT + searchTitle);
            let relatedArticles = await axios.get(WIKIPEDIA_GET_RELATED_ARTICLES_ENDPOINT + searchTitle);
            result.data.leadArticle = leadArticleResult.data;
            result.data.imdbID = imdbID;
            result.data.relatedArticles = relatedArticles.data;
            await this.setState({ searchResultWikipedia: result });
            await new Promise(r => setTimeout(r, 2000)); // TODO: remove
            await this.setState({ isLoading: false });
            await this.setState({ displayError: true });

        } catch (error) {
            console.log(error);
        }
    }

    render() {

        const {searchResult, isLoading} = this.props;
        const hasResults = (typeof searchResult !== 'undefined' && searchResult.length > 0)

        return (
            <div>
                <Spinner isLoading={isLoading} />
                {! hasResults && this.state.displayErrors &&
                    <h4>We could not find a movie with that search term. Please try using a different term...</h4>
                }
                {hasResults && <h3>IMDB Movie Database Search Results:</h3>}
                {searchResult.map((result) => (
                    <div className="resultCard" key={result.imdbID}>
                        <div className="resultData">
                        <a onClick={(event) => this.searchWikipedia(event, result.Title, result.imdbID)} href="">
                            <h3 className="articleTitle">{result.Title}</h3>
                        </a>
                        <p>Year: {result.Year}</p>
                        <p>IMDB ID: {result.imdbID}</p>
                        <p>Type: {result.Type}</p>
                        </div>
                        <Poster poster={ result.Poster }></Poster>
                        <div className="break"></div>
                        <SearchResultWikipedia isLoading={this.state.isLoading} imdbID={result.imdbID} searchResultWikipedia={ this.state.searchResultWikipedia }></SearchResultWikipedia>
                    </div>
                ))}
            </div>
        )
    }

}

export default SearchResult;