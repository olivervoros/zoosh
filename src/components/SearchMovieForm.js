import React, { Component } from 'react'
import '../App.sass';
import axios from "axios";
import SearchResult from "./SearchResult";
import { IMDB_API_ENDPOINT, slugify } from "../Helper";

class SearchMovieForm extends Component {

    constructor(props) {
        super(props);
        this.state = {movieTitle: '', searchResult: [], isLoading: false, displayError: false};

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange(event) {
        this.setState({movieTitle: event.target.value});
    }

    reset(event) {
        event.preventDefault();
        this.setState({movieTitle: '', searchResult: [], isLoading: false, displayError: false});
    }

    async handleSubmit(event) {
        event.preventDefault();
        await this.setState({ isLoading: true, displayError: false });
        const searchTitle = this.state.movieTitle;
        try {
            let result = await axios.get(IMDB_API_ENDPOINT + slugify(searchTitle));
            await new Promise(r => setTimeout(r, 2000)); // To simulate the REST API request...
            await this.setState({ isLoading: false });
            await this.setState({ searchResult: result.data.Search });
            await this.setState({ displayError: true });

        } catch (error) {
            console.log(error);
        }
    }


    render() {

        return (
            <div>
                <div>
                    <form action="#" method="post" onSubmit={this.handleSubmit}>
                        <div>
                            <label>Search for movies: </label>
                            <input className="movieTitle" value={this.state.movieTitle} onChange={this.handleChange} required id="movieTitle"
                                   name="movieTitle" type="text"/>
                        </div>
                        <p>
                            <button className="button movieSearchButton">Search</button>
                            <button onClick={(event) => this.reset(event)} className="resetButton">Reset</button>
                        </p>
                    </form>
                </div>
                <SearchResult searchResult = { this.state.searchResult } isLoading={this.state.isLoading} displayError={this.state.displayError}></SearchResult>
            </div>
        )
    }
}

export default SearchMovieForm;
