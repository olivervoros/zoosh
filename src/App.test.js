import React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import SearchMovieForm from './components/SearchMovieForm';
import Poster from './components/Poster';
import Spinner from "./components/Spinner";
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure, mount } from 'enzyme';
import SearchResult from "./components/SearchResult";
import renderer from 'react-test-renderer';

configure({adapter: new Adapter()});

test('renders movie database homepage', () => {
  const { getByText } = render(<App />);
  const companyName = getByText(/Zoosh/i);
  expect(companyName).toBeInTheDocument();
});

test('homepage_has_search_text_label', () => {
  const wrapper = mount(<App />);
  const searchText = <label>Search for movies: </label>;
  expect(wrapper.contains(searchText)).toEqual(true);
});

test('renders_search_button', () => {
  const wrapper = mount(<App />);
  expect(wrapper.contains("Search")).toEqual(true);

});

test('renders_reset_button', () => {
  const wrapper = mount(<App />);
  expect(wrapper.contains("Reset")).toEqual(true);

});

test('renders the search result component', () => {
  const movieTitle = 'Columbo';
  const isLoading = false;
  const searchResult= [];
  const props = {movieTitle, searchResult, isLoading};
  const renderedComponent = shallow(<SearchMovieForm {...props} />);
  const items = renderedComponent.find(SearchResult);
  expect(items).toHaveLength(1);
});

test('renders_the_movie_image', () => {
  const poster = 'https://m.media-amazon.com/images/M/MV5BZjEyOTE4MzMtNmMzMy00Mzc3LWJlOTQtOGJiNDE0ZmJiOTU4L2ltYWdlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg';
  const props = {poster};
  const renderedComponent = shallow(<Poster { ...props } />);
  const renderedPoster = renderedComponent.find('img');
  expect(renderedPoster).toHaveLength(1);

});

test('renders the search result data correctly', () => {
  const isLoading = false;
  const searchResult = [{"Title":"Columbo","Year":"1971â€“2003","imdbID":"tt1466074","Type":"series","Poster":"https://m.media-amazon.com/images/M/MV5BODBlYjcwNWMtZDM0OS00YzZjLTllOWYtYjg2MDM0NjQ5NjRmXkEyXkFqcGdeQXVyMjExMjk0ODk@._V1_SX300.jpg"}];
  const props = {isLoading, searchResult};
  const renderedComponent = shallow(<SearchResult {...props} />);
  const h3 = renderedComponent.find('.articleTitle');
  expect(h3.text()).toBe('Columbo');
});

test('spinner works correctly', () => {
  const isLoading = true;
  const props = {isLoading};
  const renderedComponent = shallow(<Spinner {...props} />);
  const spinner = renderedComponent.find('.spinner');
  expect(spinner).toHaveLength(1);

  const isLoading2 = false;
  const props2 = { isLoading2 };
  const renderedComponent2 = shallow(<Spinner {...props2 } />);
  const spinner2 = renderedComponent2.find('.spinner');
  expect(spinner2).toHaveLength(0);

});

test('testing search movie form renders correctly using snapshot', () => {
  const movieTitle = '';
  const isLoading = false;
  const searchResult= [];
  const props = {movieTitle, searchResult, isLoading};
  const tree = renderer.create(<SearchMovieForm {...props}></SearchMovieForm>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('testing spinner renders correctly using snapshot', () => {
  const isLoading = true;
  const props = {isLoading};
  const tree = renderer.create(<Spinner {...props}></Spinner>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('testing poster renders correctly using snapshot', () => {
  const poster = 'https://m.media-amazon.com/images/M/MV5BZjEyOTE4MzMtNmMzMy00Mzc3LWJlOTQtOGJiNDE0ZmJiOTU4L2ltYWdlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg';
  const props = {poster};
  const tree = renderer.create(<Poster {...props}></Poster>).toJSON();
  expect(tree).toMatchSnapshot();
});