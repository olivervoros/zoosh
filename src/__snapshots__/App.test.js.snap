// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing poster renders correctly using snapshot 1`] = `
<div
  className="resultPoster"
>
  <p>
    <img
      alt=""
      src="https://m.media-amazon.com/images/M/MV5BZjEyOTE4MzMtNmMzMy00Mzc3LWJlOTQtOGJiNDE0ZmJiOTU4L2ltYWdlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg"
    />
  </p>
</div>
`;

exports[`testing search movie form renders correctly using snapshot 1`] = `
<div>
  <div>
    <form
      action="#"
      method="post"
      onSubmit={[Function]}
    >
      <div>
        <label>
          Search for movies: 
        </label>
        <input
          className="movieTitle"
          id="movieTitle"
          name="movieTitle"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <p>
        <button
          className="button movieSearchButton"
        >
          Search
        </button>
        <button
          className="resetButton"
          onClick={[Function]}
        >
          Reset
        </button>
      </p>
    </form>
  </div>
  <div>
    <div />
  </div>
</div>
`;

exports[`testing spinner renders correctly using snapshot 1`] = `
<div
  className="spinner round"
/>
`;
